regression-tests/pure2-bugfix-for-optimizing-inspect.cpp2

```diff
@@ -0,0 +1,151 @@
+main: () = {
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      // Condition is `false_type` for `int`, alternative not instantiated.
+      is std::string = x.substr(0);
+      is _           = true;
+    }]]
+  }(1);
+
+  {
+    pred := :<T> (y: T) -> std::true_type = {
+      std::terminate();
+      return ();
+    };
+    [[assert Testing: inspect 1 -> bool {
+      // Condition is `true_type` for `int`, statement unconditionally executed.
+      is (pred) = true;
+      is _      = false;
+    }]]
+  }
+
+  _ = :(x) = {
+    pred := :<T> (y: T) -> bool requires std::is_same_v<T, long> = {
+      std::terminate();
+      return true;
+    };
+    [[assert Testing: inspect x -> bool {
+      // Condition is non-viable for `int`, alternative not instantiated.
+      is (pred) = x.substr(0);
+      is _      = true;
+    }]]
+  }(1);
+
+  // Exercise all built-in `is` overloads.
+
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      is std::vector          = x.substr(0);
+      is std::array           = x.substr(0);
+      is std::string          = x.substr(0);
+      is std::exception       = x.substr(0);
+      is cpp2::empty          = x.substr(0);
+      is (std::ranges::empty) = x.substr(0);
+      is ("")                 = x.substr(0);
+      is _                    = true;
+    }]]
+  }(1);
+  _ = :(x) = {
+    [[assert Testing: inspect x& -> bool {
+      is std::exception = x.substr(0);
+      is _              = true;
+    }]]
+  }(1);
+
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      is std::vector = true;
+      is _           = false;
+    }]]
+  }(:std::vector = (1));
+
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      is std::array = true;
+      is _          = false;
+    }]]
+  }(:std::array = (1));
+
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      is int = true;
+      is _   = false;
+    }]]
+  }(1);
+
+  _ = :(x) = {
+    [[assert Testing: inspect x -> bool {
+      // FIXME
+      // Using `std::ranges::subrange<std::add_pointer_t<i32>>::view_interface` fails.
+      // The alternative is elided due to the `is` being ambiguous.
+      // Like in P2392, the cases of the built-in `is` should be a chain of conditions.
+      // Using overloads to implement that is tedious and error-prone.
+      is std::ranges::view_interface<std::ranges::subrange<std::add_pointer_t<i32>>> = true;
```

Comment:
This is not a [BUG].
This is not a [SUGGESTION] (but could be taken as one).

I have thought of having the function expression's signature be `<T> (x: T)`
so that I could use `T::view_interface` in the line I'm commenting on.
But it doesn't work because it's missing the lowered Cpp1 `typename` in front of it.

Cppfront already recognizes a template function's template parameters in its function parameter list to omit wrapping a function parameter's type in `cpp2::in`.
So do you think it'd make sense to extend Cpp2 to not require `typename` in such type-only contexts?

I'm also wondering how you plan to support such disambiguating `typename` in Cpp2.

Reply:
The more I think about it,
the more I'm convinced that
"yes, this should be a thing!"

This is basically Cpp1's optional `typename`,
but for Cpp2's type-only contexts.

Reply:
Opened #531 for this.